network/src/protocol.rs

```diff
@@ -758,7 +758,29 @@ async fn statement_import_loop<Api>(
 				if let Some(producer) = producer {
 					trace!(target: "validation", "driving statement work to completion");
 
-					let work = producer.prime(api.clone()).validate();
+					let table = table.clone();
+					let gossip_handle = gossip_handle.clone();
+
+					let work = producer.prime(api.clone()).validate().map(move |res| {
+						let validated = match res {
+							Err(e) => {
+								debug!(target: "p_net", "Failed to act on statement: {}", e);
+								return
+							}
+							Ok(v) => v,
+						};
+
+						// propagate the statement.
+						let statement = crate::legacy::gossip::GossipStatement::new(
+							relay_parent,
+							match table.import_validated(validated) {
+								Some(s) => s,
+								None => return,
+							}
+						);
+
+						gossip_handle.gossip_message(topic, statement.into());
+					});
```

Comment:
This addition is not quite clear. As far as i understand it the PR mostly decouples `CandidateReceipt` into two entities (as it is described in the PR description) and introduces fixes that go along with that. But this new part which is gossiping messages looks a bit out of place. As a suggestion: if this logic was missing because of a bug and it doesn't belong here, probably it could go in a separate PR.